src/lib/cache.ts

```diff
@@ -36,3 +36,122 @@ export const useCache = <Key, Value>(
 
 	return value
 }
+
+export type EventDrivenCacheInstance<Key, Data> = {
+	subscribe: (key: Key, listener: (data: Data) => void) => () => void
+	unsubscribe: (key: Key, listener: ((data: Data) => void) | null) => void
+}
+
+export type CacheContext<Key, Data> = {
+	cache?: EventDrivenCacheInstance<Key, Data> | undefined
+	ready?: boolean
+}
+
+export const newEventDrivenCache = <Key extends string | number | symbol, Data>(
+	fn: (
+		key: Key | undefined,
+		callback: (data: Data) => void,
+		hasCache: boolean
+	) => (() => void) | undefined
+): EventDrivenCacheInstance<Key, Data> => {
+	const cache: Partial<Record<Key, Data>> = {}
+	const listeners: Partial<Record<Key, Array<(data: Data) => void>>> = {}
+	const subscriptions: Partial<Record<Key, () => void>> = {}
+
+	const unsubscribe = (key: Key, listener: ((data: Data) => void) | null) => {
+		const list = listeners[key]
+		if (!list) {
+			return
+		}
+
+		// Remove the listener from the array
+		const cleaned = []
+		for (let i = 0; i < list.length ?? 0; i++) {
+			if (list[i] === listener) {
+				listener = null
+			} else {
+				cleaned.push(list[i])
+			}
+		}
```

Comment:
Probably easer to read but untested

```suggestion
		const cleaned = list.filter(l => l !== listener)
		listener = null
```

Reply:
That would remove duplicate listeners though. We only want to remove one of them. The `listener = null` is just so we keep pushing the elements to the new list and the first `if` only ever matches once.